xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiRequestProcessor.java

```diff
@@ -19,54 +19,65 @@
+import java.util.List;
+import java.net.URL;
 
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.xwiki.configuration.ConfigurationSource;
 
-import com.xpn.xwiki.internal.XWikiCfgConfigurationSource;
+import org.xwiki.resource.internal.entity.EntityResourceActionLister;
+import org.xwiki.url.internal.standard.StandardURLConfiguration;
+import org.xwiki.url.ExtendedURL;
+
 
 /**
  * @version $Id$
  */
 public class XWikiRequestProcessor extends org.apache.struts.action.RequestProcessor
 {
+    private static final String VIEW_ACTION = "view";
+
+    private final EntityResourceActionLister entityResourceActionLister = Utils.getComponent(EntityResourceActionLister.class);
+
+    private StandardURLConfiguration configuration = Utils.getComponent(StandardURLConfiguration.class);
+
     protected static final Logger LOGGER = LoggerFactory.getLogger(XWikiRequestProcessor.class);
 
     @Override
     protected String processPath(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse)
-        throws IOException
     {
-        String result = super.processPath(httpServletRequest, httpServletResponse);
-
-        ConfigurationSource configuration =
-            Utils.getComponent(ConfigurationSource.class, XWikiCfgConfigurationSource.ROLEHINT);
-        if ("1".equals(configuration.getProperty("xwiki.virtual.usepath", "1"))) {
-            // Remove /wikiname part if the struts action is /wiki
-            if (httpServletRequest.getServletPath().equals(
-                "/" + configuration.getProperty("xwiki.virtual.usepath.servletpath", "wiki"))) {
-                int wikiNameIndex = result.indexOf("/", 1);
-                if (wikiNameIndex == -1) {
-                    result = "";
-                } else {
-                    result = result.substring(wikiNameIndex);
+        try {
+            /**
+             * Extract the action using the same method as in {@link
+             * org.xwiki.url.internal.standard.entity.AbstractEntityResourceReferenceResolver}
+             *
+             * We cannot use the EntityResourceReferenceResolver because the execution context have not been
+             * initialized here.
+             */
+            final String url = httpServletRequest.getRequestURL().toString();
+
+            final ExtendedURL extendedURL = new ExtendedURL(new URL(url), httpServletRequest.getContextPath());
+
+            final List<String> pathSegments = extendedURL.getSegments();
+
+            String action = VIEW_ACTION;
+
+            if (pathSegments.size() > 0) {
+                action = pathSegments.get(0);
+                if (this.configuration.isViewActionHidden()
+                    && !VIEW_ACTION.equals(action)
+                    && !this.entityResourceActionLister.listActions().contains(action)) {
```

Comment:
We also need to find a way to reuse the URL module more I think. You say the EC has not been initialized yet, maybe we can initialize it in a Filter before (something I've been wanting to do). But I agree we could do this in a second step since it's more involved and needs some deep thinking :)